File: src/flavor-rs/src/psp/format_2025/builder/slot_processor.rs
//! Slot processing and validation

use super::super::constants::{OP_TAR, OP_GZIP};
use super::super::defaults::{DEFAULT_FILE_PERMS};
use super::super::manifest::ManifestSlot;
use super::super::metadata::SlotMetadata;
use super::super::operations::pack_operations;
use super::super::slots::SlotDescriptor;
use super::super::checksums::{calculate_checksum, ChecksumAlgorithm};
use crate::exceptions::{FlavorError, Result};
use log::{debug, error, info, trace};
use std::fs::File;
use std::io::{BufReader, Read};
use std::path::{Path, PathBuf};
use std::time::Instant;

/// Self-referential slot markers
const SELF_REF_MARKERS: &[&str] = &["$SELF", "$LAUNCHER", "$CURRENT"];

/// Check if a slot is self-referential
///
/// A slot is self-referential if its source field contains a special marker
/// ($SELF, $LAUNCHER, or $CURRENT), indicating it references the launcher
/// itself rather than packaged data.
fn is_self_referential(source: &str) -> bool {
    SELF_REF_MARKERS.contains(&source)
}

/// Process and validate slot data
pub(super) struct SlotProcessor {
    pub(super) manifest_slots: Vec<ManifestSlot>,
    pub(super) slot_descriptors: Vec<SlotDescriptor>,
    pub(super) metadata_slots: Vec<SlotMetadata>,
    pub(super) slot_paths: Vec<PathBuf>,
}

impl SlotProcessor {
    pub(super) fn new(manifest_slots: Vec<ManifestSlot>) -> Self {
        Self {
            manifest_slots,
            slot_descriptors: Vec::new(),
            metadata_slots: Vec::new(),
            slot_paths: Vec::new(),
        }
    }

    pub(super) fn process_slots(&mut self) -> Result<()> {
        debug!("🎰 Processing {} slots", self.manifest_slots.len());
        let slots_timer = Instant::now();

        // Process slots one by one
        let num_slots = self.manifest_slots.len();
        for i in 0..num_slots {
            // Work with index to avoid borrow checker issues
            let slot = &self.manifest_slots[i];

            trace!("📖 Processing slot {}: {}", i, slot.source);

            // Validate slot number if provided
            if let Some(declared_slot) = slot.slot {
                if declared_slot as usize != i {
                    error!(
                        "❌ Critical: Slot number mismatch - expected {}, declared {} for slot '{}'",
                        i, declared_slot, slot.id
                    );
                    std::process::exit(1);
                }
            }

            // Check if this is a self-referential slot
            if is_self_referential(&slot.source) {
                info!("✨ Slot {} is self-referential ({}), skipping packaging", i, slot.source);

                // Create metadata for self-ref slot (no actual data)
                let slot_meta = SlotMetadata {
                    index: i,
                    id: slot.id.clone(),
                    source: slot.source.clone(),
                    target: slot.target.clone(),
                    size: 0,  // No data to package
                    checksum: String::new(),  // No checksum needed
                    operations: String::new(),  // No operations
                    purpose: slot.purpose.clone(),
                    lifecycle: slot.lifecycle.clone(),
                    permissions: slot.permissions.clone().or_else(|| Some(format!("{:04o}", DEFAULT_FILE_PERMS))),
                    resolution: slot.resolution.clone().or_else(|| Some("build".to_string())),
                    self_ref: Some(true),  // Mark as self-referential
                };
                self.metadata_slots.push(slot_meta);

                // Create empty descriptor (size=0, no operations)
                let descriptor = SlotDescriptor {
                    id: i as u64,
                    name_hash: 0,
                    offset: 0,  // Will be set during finalization
                    size: 0,  // No data for self-ref slot
                    original_size: 0,
                    operations: 0,  // No operations
                    checksum: 0,
                    purpose: 0,
                    lifecycle: 0,
                    priority: 0,
                    platform: 0,
                    reserved1: 0,
                    reserved2: 0,
                    permissions: 0,
                    permissions_high: 0,
                };
                self.slot_descriptors.push(descriptor);

                // Add empty path (no file to stream)
                self.slot_paths.push(PathBuf::new());

                continue;  // Skip normal processing
            }

            // Normal slot processing (non-self-ref)
            // Resolve slot path
            let slot_path = self.resolve_slot_path(&slot.source)?;

            // Calculate checksums and size
            let (file_size, sha256_checksum, adler32_checksum) = self.calculate_slot_checksums(&slot_path, i)?;

            // Create metadata entry
            let slot_meta = SlotMetadata {
                index: i,
                id: slot.id.clone(),
                source: slot.source.clone(),
                target: slot.target.clone(),
                size: file_size as i64,
                checksum: sha256_checksum,
                operations: slot.operations.clone(),
                purpose: slot.purpose.clone(),
                lifecycle: slot.lifecycle.clone(),
                permissions: slot.permissions.clone().or_else(|| Some(format!("{:04o}", DEFAULT_FILE_PERMS))),
                resolution: slot.resolution.clone().or_else(|| Some("build".to_string())),
                self_ref: None,  // Normal slot, not self-referential
            };
            self.metadata_slots.push(slot_meta);

            // Create descriptor
            let descriptor = self.create_slot_descriptor(i, slot, file_size, adler32_checksum)?;
            self.slot_descriptors.push(descriptor);

            // Store path for later streaming
            self.slot_paths.push(slot_path);
        }

        debug!(
            "✅ Processed {} slots in {:?}",
            self.manifest_slots.len(),
            slots_timer.elapsed()
        );
        Ok(())
    }

    fn resolve_slot_path(&self, source: &str) -> Result<PathBuf> {
        let slot_path = if source.contains("{workenv}") {
            // Priority: 1. FLAVOR_WORKENV_BASE env var, 2. Current working directory
            let base_dir = if let Ok(env_base) = std::env::var("FLAVOR_WORKENV_BASE") {
                info!("🔍 Using FLAVOR_WORKENV_BASE: {}", env_base);
                PathBuf::from(env_base)
            } else {
                let cwd = std::env::current_dir().map_err(|e| {
                    FlavorError::Generic(format!("Failed to get current directory: {}", e))
                })?;
                info!("🔍 No FLAVOR_WORKENV_BASE, using CWD: {}", cwd.display());
                cwd
            };
            let resolved = source.replace("{workenv}", base_dir.to_str().unwrap_or("."));
            info!(
                "📍 Resolved slot path: {} -> {} (base: {})",
                source,
                resolved,
                base_dir.display()
            );
            PathBuf::from(resolved)
        } else {
            info!("📍 Slot path has no {{workenv}}: {}", source);
            PathBuf::from(source)
        };

        info!("Attempting to open slot file at: {:?}", slot_path);
        Ok(slot_path)
    }

    fn calculate_slot_checksums(&self, slot_path: &Path, index: usize) -> Result<(u64, String, u32)> {
        let slot_file = File::open(slot_path).map_err(|e| {
            FlavorError::Generic(format!("Failed to open slot {}: {}", slot_path.display(), e))
        })?;

        let file_metadata = slot_file.metadata()?;
        let file_size = file_metadata.len();
        trace!("📊 Slot {} size: {} bytes", index, file_size);

        // Calculate SHA-256 checksum
        let checksum_timer = Instant::now();
        let mut reader = BufReader::with_capacity(8 * 1024 * 1024, slot_file);
        let sha256_checksum = calculate_checksum(&mut reader, ChecksumAlgorithm::Sha256).map_err(|e| {
            FlavorError::Generic(format!("Failed to calculate SHA256 for slot {}: {}", index, e))
        })?;

        // Calculate Adler-32 checksum
        let slot_file2 = File::open(slot_path)?;
        let mut reader2 = BufReader::with_capacity(8 * 1024 * 1024, slot_file2);
        let mut adler = adler::Adler32::new();
        let mut buffer = vec![0u8; 8 * 1024 * 1024];
        loop {
            let bytes_read = reader2.read(&mut buffer).map_err(|e| {
                FlavorError::Generic(format!("Failed to read slot {} for Adler32: {}", index, e))
            })?;
            if bytes_read == 0 {
                break;
            }
            adler.write_slice(&buffer[..bytes_read]);
        }
        let adler32_checksum = adler.checksum();

        trace!("☑️ Checksums calculated in {:?}", checksum_timer.elapsed());
        info!("Slot {}: SHA256 checksum: {}", index, sha256_checksum);
        info!("Slot {}: Adler32 checksum: {:08x}", index, adler32_checksum);

        Ok((file_size, sha256_checksum, adler32_checksum))
    }

    fn create_slot_descriptor(&self, index: usize, slot: &ManifestSlot, file_size: u64, adler_checksum: u32) -> Result<SlotDescriptor> {
        // Parse operations from comma-separated string (e.g., "tar,gzip")
        let operations = if slot.operations.is_empty() || slot.operations == "none" || slot.operations == "raw" {
            vec![]
        } else if slot.operations == "tgz" {
            // Special case for "tgz" shorthand
            vec![OP_TAR, OP_GZIP]
        } else {
            // Parse comma-separated operations
            slot.operations.split(',')
                .map(|s| s.trim())
                .filter(|s| !s.is_empty())
                .filter_map(|s| match s {
                    "tar" => Some(OP_TAR),
                    "gzip" => Some(OP_GZIP),
                    _ => {
                        log::warn!("Unknown operation: {}, skipping", s);
                        None
                    }
                })
                .collect::<Vec<u8>>()
        };

        // Map purpose string to byte value
        let purpose_value = match slot.purpose.as_str() {
            "payload" => 0,
            "runtime" => 1,
            "tool" => 2,
            _ => 0,
        };

        // Map lifecycle string to byte value
        let lifecycle_value = match slot.lifecycle.as_str() {
            "init" => 0,
            "startup" => 1,
            "runtime" => 2,
            "shutdown" => 3,
            "cache" => 4,
            "temp" => 5,
            "lazy" => 6,
            "eager" => 7,
            "dev" => 8,
            "config" => 9,
            "platform" => 10,
            _ => 2,
        };

        // Create descriptor
        let mut descriptor = SlotDescriptor::new(index as u64);
        descriptor = descriptor.with_name(&slot.id);
        descriptor.size = file_size;
        descriptor.original_size = file_size;
        descriptor.checksum = adler_checksum as u64;
        descriptor.operations = pack_operations(&operations);
        descriptor.purpose = purpose_value;
        descriptor.lifecycle = lifecycle_value;

        // Parse permissions
        let perms = if let Some(ref perm_str) = slot.permissions {
            u16::from_str_radix(perm_str.trim_start_matches('0'), 8).unwrap_or(DEFAULT_FILE_PERMS)
        } else {
            DEFAULT_FILE_PERMS
        };
        descriptor.permissions = (perms & 0xFF) as u8;
        descriptor.permissions_high = ((perms >> 8) & 0xFF) as u8;

        trace!(
            "📍 Slot {}: {} size {} bytes, checksum {:08x}",
            index,
            slot.id,
            file_size,
            adler_checksum
        );

        Ok(descriptor)
    }
}

File: src/flavor-go/pkg/psp/format_2025/slot_processor.go
package format_2025

import (
	"crypto/sha256"
	"encoding/binary"
	"fmt"
	"os"
	"strconv"
	"strings"

	"github.com/hashicorp/go-hclog"
)

// SelfRefMarker is the special marker for self-referential slots
const SelfRefMarker = "$SELF"

// computeSlotChecksum computes SHA-256 checksum truncated to first 8 bytes (uint64)
func computeSlotChecksum(data []byte) uint64 {
	hash := sha256.Sum256(data)
	return binary.LittleEndian.Uint64(hash[:8])
}

// isSelfReferential checks if a slot references the launcher itself
func isSelfReferential(source string) bool {
	return source == SelfRefMarker
}

// SlotProcessor handles slot processing for PSPF packages.
//
// This abstraction encapsulates all slot-related operations including loading,
// processing, and preparing slot data for package assembly. It aligns with the
// Rust and Python implementations for cross-language consistency in the PSPF/2025
// format.
//
// The processor handles:
// - Loading slot data from disk
// - Applying encoding/compression
// - Calculating checksums
// - Creating slot descriptors for the binary format
// - Creating slot metadata for the JSON metadata section
//
// Slot processing workflow:
// 1. Load manifest slots configuration
// 2. Process each slot (read, compress, checksum)
// 3. Create descriptors for binary slot table
// 4. Create metadata for JSON metadata section
// 5. Store compressed data for writing to package
type SlotProcessor struct {
	// Slots from manifest configuration
	manifestSlots []Slot

	// Processed slot descriptors for the package
	slotDescriptors []SlotDescriptor

	// Slot metadata for the metadata section
	metadataSlots []SlotMetadata

	// Compressed slot data to write
	slotData [][]byte

	// Logger for debug output
	logger hclog.Logger
}

// NewSlotProcessor creates a new slot processor
func NewSlotProcessor(slots []Slot, logger hclog.Logger) *SlotProcessor {
	return &SlotProcessor{
		manifestSlots:   slots,
		slotDescriptors: make([]SlotDescriptor, 0, len(slots)),
		metadataSlots:   make([]SlotMetadata, 0, len(slots)),
		slotData:        make([][]byte, 0, len(slots)),
		logger:          logger,
	}
}

// ProcessSlots processes all slots from the manifest.
//
// This method iterates through all configured slots, loading their data,
// applying any specified encoding, and preparing both the binary descriptors
// and JSON metadata. Each slot is validated for required fields and processed
// according to its configuration.
//
// Returns an error if any slot fails to process.
func (sp *SlotProcessor) ProcessSlots() error {
	sp.logger.Info("📦 Processing slot metadata", "count", len(sp.manifestSlots))
	sp.logger.Debug("🔍 Slot processing details", "alignment", SlotAlignment, "descriptor_size", SlotDescriptorSize)

	for i, slot := range sp.manifestSlots {
		if err := sp.processSlot(i, &slot); err != nil {
			return fmt.Errorf("failed to process slot %d: %w", i, err)
		}
	}

	return nil
}

// mapPurposeToUint8 maps purpose string to uint8 value for binary format
func mapPurposeToUint8(purpose string) uint8 {
	switch purpose {
	case "payload":
		return 0
	case "runtime":
		return 1
	case "tool":
		return 2
	default:
		return 0 // default to payload
	}
}

// mapLifecycleToUint8 maps lifecycle string to uint8 value for binary format
func mapLifecycleToUint8(lifecycle string) uint8 {
	switch lifecycle {
	// Timing-based
	case "init":
		return 0
	case "startup":
		return 1
	case "runtime":
		return 2
	case "shutdown":
		return 3
	// Retention-based
	case "cache":
		return 4
	case "temp":
		return 5
	// Access-based
	case "lazy":
		return 6
	case "eager":
		return 7
	// Environment-based
	case "dev":
		return 8
	case "config":
		return 9
	case "platform":
		return 10
	default:
		return 2 // default to runtime
	}
}

// parsePermissions parses permission string (e.g., "0755") to uint16
func parsePermissions(permStr string) uint16 {
	if permStr == "" {
		return uint16(FilePerms)
	}

	// Parse octal string (e.g., "0755" -> 0o755)
	cleaned := strings.TrimPrefix(permStr, "0")
	if parsed, err := strconv.ParseUint(cleaned, 8, 16); err == nil {
		return uint16(parsed)
	}

	return uint16(FilePerms) // fallback to default
}

// hashSlotName is defined in builder.go and reused here

// processSlot processes a single slot.
//
// This method handles the complete processing of a single slot:
// - Validates required fields (id, source, target)
// - Sets defaults for optional fields
// - Loads and encodes the slot data
// - Calculates checksums
// - Creates slot descriptor for binary format
// - Creates slot metadata for JSON format
//
// Args:
//
//	index: The slot index (0-based)
//	slot: The slot configuration from the manifest
//
// Returns an error if the slot cannot be processed.
func (sp *SlotProcessor) processSlot(index int, slot *Slot) error {
	// Validate required fields
	if slot.ID == "" {
		return fmt.Errorf("slot %d missing required 'id' field", index)
	}
	if slot.Source == "" {
		return fmt.Errorf("slot %d missing required 'source' field (id: %s)", index, slot.ID)
	}
	if slot.Target == "" {
		return fmt.Errorf("slot %d missing required 'target' field (id: %s)", index, slot.ID)
	}

	// Set defaults
	if slot.Resolution == "" {
		slot.Resolution = "build"
	}
	if slot.Permissions == "" {
		slot.Permissions = fmt.Sprintf("%04o", FilePerms)
	}

	// Validate slot number if provided
	if slot.Slot != nil && *slot.Slot != index {
		return fmt.Errorf("slot number mismatch: expected %d, declared %d (id: %s)",
			index, *slot.Slot, slot.ID)
	}

	sp.logger.Debug("📂 Processing slot", "index", index, "id", slot.ID,
		"source", slot.Source, "target", slot.Target)

	// Check if this is a self-referential slot
	if isSelfReferential(slot.Source) {
		sp.logger.Info("✨ Slot is self-referential, skipping packaging",
			"index", index, "source", slot.Source)

		// Create metadata for self-ref slot (no actual data)
		selfRefTrue := true
		slotMeta := SlotMetadata{
			Slot:        index,
			ID:          slot.ID,
			Source:      slot.Source,
			Target:      slot.Target,
			Size:        0,  // No data to package
			Checksum:    "", // No checksum needed
			Operations:  "", // No operations
			Purpose:     slot.Purpose,
			Lifecycle:   slot.Lifecycle,
			Resolution:  slot.Resolution,
			Permissions: slot.Permissions,
			SelfRef:     &selfRefTrue, // Mark as self-referential
		}

		// Create empty descriptor (size=0, no operations)
		descriptor := SlotDescriptor{
			ID:              uint64(index),
			NameHash:        HashName(slot.Target),
			Offset:          0, // Will be set during finalization
			Size:            0, // No data for self-ref slot
			OriginalSize:    0,
			Operations:      0, // No operations
			Checksum:        0,
			Purpose:         mapPurposeToUint8(slot.Purpose),
			Lifecycle:       mapLifecycleToUint8(slot.Lifecycle),
			Priority:        128,
			Platform:        0,
			Reserved1:       0,
			Reserved2:       0,
			Permissions:     uint8(parsePermissions(slot.Permissions) & 0xFF),
			PermissionsHigh: uint8(parsePermissions(slot.Permissions) >> 8),
		}

		// Store processed data
		sp.metadataSlots = append(sp.metadataSlots, slotMeta)
		sp.slotDescriptors = append(sp.slotDescriptors, descriptor)
		sp.slotData = append(sp.slotData, []byte{}) // Empty data

		sp.logger.Debug("✅ Self-referential slot processed", "index", index, "id", slot.ID)
		return nil
	}

	// Normal slot processing - read and process slot data
	slotData, compressed, _, err := sp.loadSlotData(slot)
	if err != nil {
		return fmt.Errorf("failed to load slot data: %w", err)
	}

	// Calculate checksum of compressed data
	checksumData := sha256.Sum256(compressed)
	checksumStr := fmt.Sprintf("sha256:%x", checksumData)

	// Create slot metadata
	slotMeta := SlotMetadata{
		Slot:        index,
		ID:          slot.ID,
		Source:      slot.Source,
		Target:      slot.Target,
		Size:        int64(len(slotData)),
		Checksum:    checksumStr,
		Operations:  slot.Operations,
		Purpose:     slot.Purpose,
		Lifecycle:   slot.Lifecycle,
		Resolution:  slot.Resolution,
		Permissions: slot.Permissions,
		SelfRef:     nil, // Normal slot, not self-referential
	}

	// Determine operations from operations field
	var operations uint64
	switch slot.Operations {
	case "gzip":
		operations = PackOperations([]uint8{OP_GZIP})
	case "tgz", "tar.gz":
		operations = PackOperations([]uint8{OP_TAR, OP_GZIP})
	case "tar":
		operations = PackOperations([]uint8{OP_TAR})
	default:
		operations = 0 // No operations (raw)
	}

	// Create slot descriptor with new structure
	descriptor := SlotDescriptor{
		ID:              uint64(index),
		NameHash:        HashName(slot.Target),
		Offset:          0, // Will be set during write phase
		Size:            uint64(len(compressed)),
		OriginalSize:    uint64(len(slotData)),
		Operations:      operations,
		Checksum:        computeSlotChecksum(compressed), // SHA-256 first 8 bytes
		Purpose:         mapPurposeToUint8(slot.Purpose),
		Lifecycle:       mapLifecycleToUint8(slot.Lifecycle),
		Priority:        128, // normal priority
		Platform:        0,   // all platforms
		Reserved1:       0,
		Reserved2:       0,
		Permissions:     uint8(parsePermissions(slot.Permissions) & 0xFF),
		PermissionsHigh: uint8(parsePermissions(slot.Permissions) >> 8),
	}

	// Store processed data
	sp.metadataSlots = append(sp.metadataSlots, slotMeta)
	sp.slotDescriptors = append(sp.slotDescriptors, descriptor)
	sp.slotData = append(sp.slotData, compressed)

	sp.logger.Debug("✅ Slot processed", "index", index, "id", slot.ID,
		"compressed_size", len(compressed), "original_size", len(slotData))

	return nil
}

// loadSlotData loads and processes slot data based on encoding.
//
// This method reads the slot data from disk and applies any specified
// encoding. It supports path resolution with {workenv} placeholder and
// various codec formats (gzip, tar, tar.gz, none).
//
// Args:
//
//	slot: The slot configuration containing source path and encoding
//
// Returns:
//   - Original uncompressed data
//   - Compressed/encoded data
//   - Encoding method constant for the binary format
//   - Error if the data cannot be loaded
func (sp *SlotProcessor) loadSlotData(slot *Slot) ([]byte, []byte, uint8, error) {
	// Resolve {workenv} placeholder
	slotPath := slot.Source
	if strings.Contains(slotPath, "{workenv}") {
		baseDir := os.Getenv("FLAVOR_WORKENV_BASE")
		if baseDir == "" {
			baseDir, _ = os.Getwd()
		}
		slotPath = strings.ReplaceAll(slotPath, "{workenv}", baseDir)
		sp.logger.Debug("📍 Resolved path", "original", slot.Source,
			"resolved", slotPath, "base", baseDir)
	}

	// Read slot data
	slotData, err := os.ReadFile(slotPath)
	if err != nil {
		return nil, nil, 0, fmt.Errorf("failed to read slot from %s: %w", slotPath, err)
	}

	sp.logger.Debug("📊 Slot size", "original", len(slotData), "operations", slot.Operations)

	// For now, we don't actually compress here - that's handled elsewhere
	// This function just validates and prepares the data
	compressed := slotData

	return slotData, compressed, 0, nil
}

// GetDescriptors returns the processed slot descriptors
func (sp *SlotProcessor) GetDescriptors() []SlotDescriptor {
	return sp.slotDescriptors
}

// GetMetadata returns the processed slot metadata
func (sp *SlotProcessor) GetMetadata() []SlotMetadata {
	return sp.metadataSlots
}

// GetSlotData returns the compressed slot data
func (sp *SlotProcessor) GetSlotData() [][]byte {
	return sp.slotData
}
